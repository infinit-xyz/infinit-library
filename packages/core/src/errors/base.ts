// ref: https://github.com/wevm/viem/blob/main/src/errors/base.ts
import { name as coreName, version as coreVersion } from 'package.json'

//TODO: add error code: eg. HH01: ...

export type ProtocolModuleVersionDetail = {
  name: string
  version: string
}

type BaseErrorParameters = {
  cause?: BaseError | Error | undefined
  details?: string | undefined
  name?: string | undefined
}

export type BaseErrorType = BaseError & { name: 'BaseError' }
export class BaseError extends Error {
  details: string
  metaMessages?: string[] | undefined
  shortMessage: string
  version: string

  override name = 'BaseError'

  constructor(shortMessage: string, args: BaseErrorParameters = {}) {
    const details = (() => {
      if (args.cause instanceof BaseError) return args.cause.details
      if (args.cause?.message) return args.cause.message
      return args.details!
    })()

    const message = [
      shortMessage || 'An error occurred.',
      '',
      ...(details ? [`Details: ${details}`] : []),
      `${coreName}: ${coreVersion}`,
    ].join('\n')

    super(message, args.cause ? { cause: args.cause } : undefined)

    this.details = details
    this.name = args.name ?? this.name
    this.shortMessage = shortMessage
    this.version = coreVersion
  }
}
