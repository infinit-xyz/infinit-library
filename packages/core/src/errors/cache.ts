import { BaseError } from '@errors/base'

export class FoundInvalidCachedTxError extends BaseError {
  constructor() {
    super(`Found a successful Tx after a failed Tx, please contract support`, {
      name: 'FoundInvalidCachedTxError',
    })
  }
}

export class IncorrectCacheError extends BaseError {
  constructor(reason = 'None') {
    super([`Found a cache but the cache is incorrect`, `Reason: ${reason}`].join('\n'), {
      name: 'IncorrectCacheError',
    })
  }
}
