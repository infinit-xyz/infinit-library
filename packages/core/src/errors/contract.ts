import { BaseError } from '@/errors/base.ts'

export type ContractValidateErrorType = ContractValidateError & {
  name: 'ContractValidateError'
}
export class ContractValidateError extends BaseError {
  constructor(extraMsg: string = '') {
    super(extraMsg, {
      name: 'ContractValidateError',
    })
  }
}
