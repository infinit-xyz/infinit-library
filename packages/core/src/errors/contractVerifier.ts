import { Address } from 'viem'

import { BaseError } from '@errors/base.ts'

export class VerifyContractError extends BaseError {
  constructor(contractName: string, contractAddress: Address, message?: string) {
    const errorDetails = [`Failed to verify contract ${contractName} on the block explorer at contract address: ${contractAddress}`]

    if (message) {
      errorDetails.push(`Message: ${message}`)
    }

    super(errorDetails.join('\n'), {
      name: VerifyContractError.name,
    })
  }
}
