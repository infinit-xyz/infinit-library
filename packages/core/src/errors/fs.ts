import { BaseError } from '@errors/base.ts'

export type DirectoryNotFoundErrorType = DirectoryNotFoundError & {
  name: 'DirectoryNotFoundError'
}
export class DirectoryNotFoundError extends BaseError {
  constructor(path: string) {
    super(`${path} not found`, {
      name: 'DirectoryNotFoundError',
    })
  }
}
