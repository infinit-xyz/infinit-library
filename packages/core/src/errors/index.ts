export * from './cache'
export * from './contract'
export * from './fs'
export * from './infinitWallet'
export * from './transaction'
export * from './validate'
