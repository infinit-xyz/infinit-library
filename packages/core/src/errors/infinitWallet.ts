import { BaseError } from '@/errors/base.ts'

export type InfinitWalletNotFoundErrorType = InfinitWalletNotFoundError & {
  name: 'InfinitWalletNotFoundError'
}
export class InfinitWalletNotFoundError extends BaseError {
  constructor() {
    super(`InfinitWallet not found`, {
      name: 'InfinitWalletNotFoundError',
    })
  }
}
