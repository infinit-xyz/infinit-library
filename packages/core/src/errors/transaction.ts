import { Hex } from 'viem'

import { BaseError } from '@/errors/base.ts'

export type TransactionErrorType = TransactionError & {
  name: 'TransactionError'
}
export class TransactionError extends BaseError {
  constructor(txHash: Hex) {
    super(`Transaction ${txHash} reverted`, {
      name: 'TransactionError',
    })
  }
}

export type ContractNotFoundErrorType = ContractNotFoundError & {
  name: 'ContractNotFoundError'
}
export class ContractNotFoundError extends BaseError {
  constructor(txHash: Hex, contractName: string = '') {
    super(`Contract ${contractName} not found in ${txHash}`, {
      name: 'ContractNotFoundError',
    })
  }
}

export type TxNotFoundErrorType = TxNotFoundError & {
  name: 'TxNotFoundError'
}
export class TxNotFoundError extends BaseError {
  constructor() {
    super('Transaction not found', {
      name: 'TxNotFoundError',
    })
  }
}
