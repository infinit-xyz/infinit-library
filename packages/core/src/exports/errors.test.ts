import * as errors from './errors'
import { expect, test } from 'vitest'

test('exports - errors: should match snapshot', () => {
  expect(Object.keys(errors)).toMatchInlineSnapshot(`
    [
      "FoundInvalidCachedTxError",
      "IncorrectCacheError",
      "ContractValidateError",
      "DirectoryNotFoundError",
      "InfinitWalletNotFoundError",
      "TransactionError",
      "ContractNotFoundError",
      "TxNotFoundError",
      "ValidateInputValueError",
      "ValueNotFoundError",
    ]
  `)
})
