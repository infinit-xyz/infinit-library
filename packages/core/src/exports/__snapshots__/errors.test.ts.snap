// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`exports - errors: should match snapshot 1`] = `
[
  "FoundInvalidCachedTxError",
  "IncorrectCacheError",
  "ContractValidateError",
  "DirectoryNotFoundError",
  "InfinitWalletNotFoundError",
  "TransactionError",
  "ContractNotFoundError",
  "TxNotFoundError",
  "InvalidEventError",
  "ValidateInputValueError",
  "ValidateLengthError",
  "ValidateInputZeroAddressError",
  "ValueNotFoundError",
]
`;
